src/dbnode/storage/index/block.go

```diff
@@ -94,26 +162,45 @@ func NewBlock(
 	md namespace.Metadata,
 	opts Options,
 ) (Block, error) {
-	var (
-		blockSize = md.Options().IndexOptions().BlockSize()
-	)
+	docsPool := opts.DocumentArrayPool()
 
-	// FOLLOWUP(prateek): use this to track segments when we have multiple segments in a Block.
-	postingsOffset := postings.ID(0)
-	seg, err := mem.NewSegment(postingsOffset, opts.MemSegmentOptions())
+	foregroundCompactor, err := compaction.NewCompactor(docsPool,
+		documentArrayPoolCapacity,
+		opts.SegmentBuilderOptions(),
+		opts.FSTSegmentOptions())
 	if err != nil {
 		return nil, err
 	}
 
-	b := &block{
-		state:         blockStateOpen,
-		activeSegment: seg,
+	backgroundCompactor, err := compaction.NewCompactor(docsPool,
+		documentArrayPoolCapacity,
+		opts.SegmentBuilderOptions(),
+		opts.FSTSegmentOptions())
+	if err != nil {
+		return nil, err
+	}
 
-		startTime: startTime,
-		endTime:   startTime.Add(blockSize),
-		blockSize: blockSize,
-		opts:      opts,
-		nsMD:      md,
+	segmentBuilder, err := builder.NewBuilderFromDocuments(opts.SegmentBuilderOptions())
+	if err != nil {
+		return nil, err
+	}
+
+	blockSize := md.Options().IndexOptions().BlockSize()
+	iopts := opts.InstrumentOptions()
+	b := &block{
+		state:               blockStateOpen,
+		segmentBuilder:      segmentBuilder,
+		startTime:           startTime,
```

Comment:
presumably this was truncated already and such? might be nice to rename it to `blockStart` if so